pvlib/ivtools.py

```diff
@@ -0,0 +1,304 @@
+    gamma_pmp : float
+        Temperature coefficient of power at maximum point point [%/C]
+    cells_in_series : int
+        Number of cells in series
+    temp_ref : float, default 25C
+        Reference temperature condition
+
+    Returns
+    -------
+    tuple of the following elements:
+
+        a_ref : float
+            The product of the usual diode ideality factor ``n`` (unitless),
+            number of cells in series ``Ns``, and cell thermal voltage at
+            reference conditions [V]
+
+        I_L_ref : float
+            The light-generated current (or photocurrent) at reference
+            conditions [A]
+
+        I_o_ref : float
+            The dark or diode reverse saturation current at reference
+            conditions [A]
+
+        R_sh_ref : float
+            The shunt resistance at reference conditions, in ohms.
+
+        R_s : float
+            The series resistance at reference conditions, in ohms.
+
+        Adjust : float
+            The adjustment to the temperature coefficient for short circuit
+            current, in percent.
+
+    Raises:
+        ImportError if NREL-PySAM is not installed
+
+    Notes
+    -----
+    Inputs ``v_mp``, ``v_oc``, ``i_mp`` and ``i_sc`` are assumed to be from a
+    single IV curve at constant irradiance and cell temperature. Irradiance is
+    not explicitly used by the fitting procedure. The irradiance level at which
+    the input IV curve is determined and the specified cell temperature
+    ``Tref`` are the reference conditions for the output parameters
+    ``I_L_ref``, ``I_o_ref``, ``R_sh_ref``, ``R_s`` and ``Adjust``.
+
+    If the fitting fails, returns NaN in each parameter.
+
+    References
+    ----------
+    [1] A. Dobos, "An Improved Coefficient Calculator for the California
+    Energy Commission 6 Parameter Photovoltaic Module Model", Journal of
+    Solar Energy Engineering, vol 134, 2012.
+    '''
+
+    try:
+        from PySAM import PySSC
+    except ImportError as e:
+        raise("Requires NREL's PySAM package at "
+              "https://pypi.org/project/NREL-PySAM/.") from e
+
+    datadict = {'tech_model': '6parsolve', 'financial_model': 'none',
+                'celltype': celltype, 'Vmp': v_mp,
+                'Imp': i_mp, 'Voc': v_oc, 'Isc': i_sc, 'alpha_isc': alpha_sc,
+                'beta_voc': beta_voc, 'gamma_pmp': gamma_pmp,
+                'Nser': cells_in_series, 'Tref': temp_ref}
+
+    result = PySSC.ssc_sim_from_dict(datadict)
+    if result['cmod_success'] == 1:
+        a_ref = result['a']
+        I_L_ref = result['Il']
+        I_o_ref = result['Io']
+        R_s = result['Rs']
+        R_sh_ref = result['Rsh']
+        Adjust = result['Adj']
+    else:
+        a_ref = np.nan
+        I_L_ref = np.nan
+        I_o_ref = np.nan
+        R_s = np.nan
+        R_sh_ref = np.nan
+        Adjust = np.nan
+    return I_L_ref, I_o_ref, R_sh_ref, R_s, a_ref, Adjust
```

Comment:
Is it really worth the effort to copy and rename the result?  The names don't get returned--only the values.

Reply:
Agree that the `return (tuple of variables)` is more compact and easier to read. I'd prefer to have the `if/then` structure with the single `return` in case we come up with a better fallback than returning `np.nan` when the SAM function doesn't work (e.g., SAM might expose some numerical options and we could retry the fitting).  There aren't options available now, that I know, but I like having the structure in place in case they become available.

FYI - `return [result[p] …` returns a list, `return result['Il'], result[…` returns a tuple.  They are a little different, I don't really understand why to prefer one over the other.  I see tuples used to pass values from functions, so I'm using one here.

Reply:
AFAIK it's best practice to use a tuple as a keyword default input to a function because tuples are immutable,  whereas a list can change,  but I don't see why a list couldn't be the output.

From [Common Gotchas](https://docs.python-guide.org/writing/gotchas/):

>Seemingly the most common surprise new Python programmers encounter is Python’s treatment of mutable default arguments in function definitions.

Reply:
I think sticking with a tuple as the return value is much more consistent with the current API.

I note the alternative "required keyword input arguments" and "dictionary of outputs" approach that I currently prefer for Python3 APIs: https://github.com/pvlib/pvlib-python/issues/660#issuecomment-464187557, which is much more forgiving w.r.t. ordering and is still convenient for call chaining, but is somewhat less performant (I think). Also remember that default values become a breakable part of the API.

Reply:
Agree that it is better to return tuples. I also like @adriesse's compact notation suggestion. You can do both if you wrap the list/generator comprehension in `tuple`: `tuple((result[p] for p in ['Il', 'Io', 'Rsh', 'Rs', 'a', 'Adj']))`. 

Does PySSC return a useful error message? If so, perhaps better to raise an exception to the user with the error message than return nan.

Reply:
PySSC prints messages to the screen. The message for failing to extract the parameters is not helpful. It looks to me that the [core functions](https://github.com/NREL/ssc/blob/1ee0dfdd9a46ce3ba80560bad650902c699f2780/shared/6par_solve.h) in the SSC check the solution for reasonable values of each parameter and for convergence to the data (look at the sanity() function if you are curious). An error code is set but no message is generated. Its a bit opaque in its current state.

Reply:
We could still set an exception if `nans` are being returned, but that would interrupt execution if someone was processing several input sets.

Reply:
> that would interrupt execution if someone was processing several input sets.

This is typically the recommended approach in Python. Let the user handle the exception unless you can safely assume what the right solution is for the vast majority of workflows. In the present state, the user would need to check that all results are not-nan to be sure that every input set succeeded. In this case, I'm ok with whatever the other reviewers of this PR prefer.

Reply:
The use case I have in mind is recreating the CEC model parameter database from the CEC's module characteristics data. There's 2000+ lines to process. An Exception that stops processing on the first line where the SAM algorithms failed to converge would be a bit frustrating, in my opinion.

@mikofski @adriesse what do you think?

Reply:
I'm relying more an more on exceptions in my work and it doesn't seem like a bad choice here.  `try/except` will let you keep processing and let you choose whether to store nans, skip that module, log the error, or whatever.

Reply:
While I think @cwhanse’s use case could still work with these exceptions, I personally like to reserve exceptions for truly exceptional cases, whereas here a numerical method failed “gracefully” (as opposed to exceptionally). That said, asking the user to search for nan’s to determine that things failed gracefully seems less than desirable.

If we go the Exception route, then I think we should define a custom pvlib exception that can be used throughout for this situation to unify the API. (Defining the scope “this situation” may be hard though!)

Reply:
There was a similar discussion on the [vectorize newton PR in SciPy](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.newton.html), and they opted to display a `RuntimeWarning`, unless the optional arg `disp=True` in which case raise [`RuntimeError`](../../../scipy/scipy/blob/master/scipy/optimize/zeros.py#L361). However, the use of the keyword `disp` was considered awkward, and only maintained  for backwards compatibility. The [statsmodels author's opinion was a return flag](../../../scipy/scipy/pull/9568#issuecomment-444168546), similar to Cliff's NaNs idea. Someone else thought it should fall back to another algorithm. Etc. Lot's of competing opinions, no right answer.

My personal opinion is to favor exceptions. I think in MATLAB and other languages we learn to fear exceptions, but Python tends to embrace them. I think the argument for raising an exception is that it informs the user and let's them quickly decide how to handle it. It also avoids a lot of type-checking and null-pointer testing which isn't necessary in a memory-managed language. If I were running an example with 2000 points, I would want to know right away that it wasn't working, so I could fix it and move on. I think I would be more frustrated if the fitting continued without telling me anything, and then I ended up with an array full of NaNs. Aaaarrrrgggghhhh!